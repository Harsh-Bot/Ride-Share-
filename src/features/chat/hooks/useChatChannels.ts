import { useQuery } from '@tanstack/react-query';

export const useChatChannels = () => {
  return useQuery({
    queryKey: ['chat-channels'],
    queryFn: async () => {
      // TODO: Integrate with Firestore/third-party chat SDK for ride-specific conversations
      return [] as Array<unknown>;
    }
  });
};
