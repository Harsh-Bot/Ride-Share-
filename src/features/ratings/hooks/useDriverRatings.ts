import { useQuery } from '@tanstack/react-query';

export const useDriverRatings = (driverId: string) => {
  return useQuery({
    queryKey: ['driver-ratings', driverId],
    queryFn: async () => {
      // TODO: Retrieve aggregated rating insights and recent feedback entries
      return {
        average: 0,
        count: 0
      };
    }
  });
};
