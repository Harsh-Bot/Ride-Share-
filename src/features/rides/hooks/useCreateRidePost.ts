import { useEffect, useMemo, useRef, useState } from 'react';
import { FirebaseError } from 'firebase/app';
import type { PostRideSubmitPayload } from '../types/postRide';
import type {
  RidePostClientSnapshot,
  CreateRidePostFn,
  SubscribeToRidePostFn
} from '../api/ridePosts';
import {
  createRidePost as firestoreCreateRidePost,
  subscribeToRidePost as firestoreSubscribeToRidePost,
  mapFirebaseError
} from '../api/ridePosts';

export type PostRideClientStatus = 'idle' | 'posting' | 'live' | 'error' | 'queued';

export type UseCreateRidePostOptions = {
  driverId: string;
  createRidePostFn?: CreateRidePostFn;
  subscribeToRidePostFn?: SubscribeToRidePostFn;
};

export type UseCreateRidePostResult = {
  status: PostRideClientStatus;
  activePost: RidePostClientSnapshot | null;
  error: string | null;
  pendingCount: number;
  postRide: (payload: PostRideSubmitPayload) => Promise<void>;
  retryPending: () => Promise<void>;
  clearError: () => void;
};

type RidePostControllerState = {
  status: PostRideClientStatus;
  activePost: RidePostClientSnapshot | null;
  error: string | null;
  pendingCount: number;
};

type PendingPost = {
  driverId: string;
  payload: PostRideSubmitPayload;
};

const formatError = (error: FirebaseError | null, fallback: string) => {
  if (!error) {
    return fallback;
  }
  switch (error.code) {
    case 'permission-denied':
      return 'You do not have permission to post this ride.';
    case 'unavailable':
      return 'Network unavailable. Ride post queued for retry.';
    default:
      return error.message || fallback;
  }
};

type RidePostControllerDeps = {
  driverId: string;
  createRidePostFn: CreateRidePostFn;
  subscribeToRidePostFn: SubscribeToRidePostFn;
};

export class RidePostController {
  private state: RidePostControllerState = {
    status: 'idle',
    activePost: null,
    error: null,
    pendingCount: 0
  };
  private listeners = new Set<(state: RidePostControllerState) => void>();
  private pendingQueue: PendingPost[] = [];
  private unsubscribe: (() => void) | null = null;
  private hasResolvedSnapshot = false;
  private driverId: string;

  constructor(private readonly deps: RidePostControllerDeps) {
    this.driverId = deps.driverId;
  }

  getState() {
    return this.state;
  }

  subscribe(listener: (state: RidePostControllerState) => void) {
    this.listeners.add(listener);
    listener(this.state);
    return () => {
      this.listeners.delete(listener);
    };
  }

  setDriverId(driverId: string) {
    this.driverId = driverId;
  }

  dispose() {
    this.clearSubscription();
    this.listeners.clear();
    this.pendingQueue = [];
  }

  async postRide(payload: PostRideSubmitPayload) {
    if (this.state.status === 'posting') {
      return;
    }

    this.clearSubscription();
    this.updateState({ status: 'posting', error: null });

    try {
      const { postId, snapshot } = await this.deps.createRidePostFn({
        driverId: this.driverId,
        payload
      });

      this.handleSubscribe(postId, snapshot);
    } catch (unknownError) {
      this.clearSubscription();
      const firebaseError = mapFirebaseError(unknownError);

      if (firebaseError?.code === 'unavailable') {
        this.pendingQueue.push({ driverId: this.driverId, payload });
        this.updateState({
          status: 'queued',
          error: formatError(firebaseError, 'Network unavailable. Ride queued.'),
          activePost: null,
          pendingCount: this.pendingQueue.length
        });
        return;
      }

      this.updateState({
        status: 'error',
        error: formatError(firebaseError, 'Unable to post ride.'),
        activePost: null
      });
    }
  }

  async retryPending() {
    if (this.state.status === 'posting' || this.pendingQueue.length === 0) {
      return;
    }

    const next = this.pendingQueue.shift();
    this.updateState({ pendingCount: this.pendingQueue.length });

    if (!next) {
      return;
    }

    const originalDriverId = this.driverId;
    this.setDriverId(next.driverId);
    await this.postRide(next.payload);
    this.setDriverId(originalDriverId);
  }

  clearError() {
    if (!this.state.error) {
      return;
    }
    const nextStatus = this.state.status === 'error' && !this.state.activePost ? 'idle' : this.state.status;
    this.updateState({ error: null, status: nextStatus });
  }

  private updateState(patch: Partial<RidePostControllerState>) {
    this.state = { ...this.state, ...patch };
    for (const listener of this.listeners) {
      listener(this.state);
    }
  }

  private clearSubscription() {
    if (this.unsubscribe) {
      this.unsubscribe();
    }
    this.unsubscribe = null;
    this.hasResolvedSnapshot = false;
  }

  private handleSubscribe(postId: string, snapshot: RidePostClientSnapshot) {
    this.hasResolvedSnapshot = false;
    this.unsubscribe = this.deps.subscribeToRidePostFn({
      postId,
      onSnapshot: (updatedSnapshot) => {
        this.hasResolvedSnapshot = true;
        this.updateState({
          activePost: updatedSnapshot,
          status: 'live',
          pendingCount: this.pendingQueue.length
        });
      },
      onError: (subscriptionError) => {
        const firebaseError = mapFirebaseError(subscriptionError);
        this.updateState({
          status: 'error',
          error: formatError(firebaseError, 'Real-time updates failed.'),
          activePost: null,
          pendingCount: this.pendingQueue.length
        });
        this.clearSubscription();
      }
    });

    if (!this.hasResolvedSnapshot) {
      this.updateState({
        activePost: snapshot,
        status: 'posting',
        pendingCount: this.pendingQueue.length
      });
    }
  }
}

export const useCreateRidePost = ({
  driverId,
  createRidePostFn = firestoreCreateRidePost,
  subscribeToRidePostFn = firestoreSubscribeToRidePost
}: UseCreateRidePostOptions): UseCreateRidePostResult => {
  const controllerRef = useRef<RidePostController | null>(null);

  if (!controllerRef.current) {
    controllerRef.current = new RidePostController({
      driverId,
      createRidePostFn,
      subscribeToRidePostFn
    });
  }

  const controller = controllerRef.current!;
  const [state, setState] = useState<RidePostControllerState>(controller.getState());

  useEffect(() => {
    controller.setDriverId(driverId);
  }, [controller, driverId]);

  useEffect(() => {
    const unsubscribe = controller.subscribe(setState);
    return () => {
      unsubscribe();
      controller.dispose();
      controllerRef.current = null;
    };
  }, [controller]);

  return useMemo(() => ({
    status: state.status,
    activePost: state.activePost,
    error: state.error,
    pendingCount: state.pendingCount,
    postRide: (payload: PostRideSubmitPayload) => controller.postRide(payload),
    retryPending: () => controller.retryPending(),
    clearError: () => controller.clearError()
  }), [state, controller]);
};
