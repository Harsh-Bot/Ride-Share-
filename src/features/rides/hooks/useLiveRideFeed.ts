import { useQuery } from '@tanstack/react-query';

export const useLiveRideFeed = () => {
  return useQuery({
    queryKey: ['live-rides'],
    queryFn: async () => {
      // TODO: Replace with Firestore/Functions powered live ride feed
      return [] as Array<unknown>;
    }
  });
};
