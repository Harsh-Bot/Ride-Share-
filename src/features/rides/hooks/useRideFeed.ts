import { useEffect, useMemo, useRef, useState } from 'react';
import { collection, onSnapshot as fsOnSnapshot, query, getDocs } from 'firebase/firestore';
import { getFirestoreDb } from '../../../services/firebase';

export type RideFeedItem = {
  id: string;
  destinationCampus: string;
  seatsAvailable: number;
  status: string;
};

export type RideFeedState = {
  items: Array<RideFeedItem & { isStale: boolean }>;
  offline: boolean;
  refreshing: boolean;
  lastServerSyncAt: number | null;
};

export type SubscribeToRideFeedFn = (options: {
  onSnapshot: (items: RideFeedItem[], fromServer: boolean) => void;
  onError: (error: unknown) => void;
  geohashPrefixes?: string[];
}) => () => void;

export type RefreshRideFeedFn = () => Promise<RideFeedItem[]>;

const FIVE_MIN = 5 * 60_000;

type RideFeedControllerDeps = {
  subscribe: SubscribeToRideFeedFn;
  refreshFn: RefreshRideFeedFn;
  now?: () => number;
};

export class RideFeedController {
  private state: RideFeedState = { items: [], offline: false, refreshing: false, lastServerSyncAt: null };
  private listeners = new Set<(state: RideFeedState) => void>();
  private unsubscribe: (() => void) | null = null;
  private now: () => number;

  constructor(private readonly deps: RideFeedControllerDeps) {
    this.now = deps.now ?? Date.now;
  }

  getState() { return this.state; }

  subscribe(listener: (state: RideFeedState) => void) {
    this.listeners.add(listener);
    listener(this.state);
    if (!this.unsubscribe) {
      this.unsubscribe = this.deps.subscribe({
        onSnapshot: (items, fromServer) => {
          const ts = fromServer ? this.now() : this.state.lastServerSyncAt;
          this.state = {
            ...this.state,
            items: this.tagStale(items, ts),
            offline: !fromServer && (this.state.offline || true),
            lastServerSyncAt: ts ?? null
          };
          this.emit();
        },
        onError: (_e) => {
          this.state = { ...this.state, offline: true };
          this.emit();
        }
      });
    }
    return () => {
      this.listeners.delete(listener);
    };
  }

  dispose() {
    if (this.unsubscribe) this.unsubscribe();
    this.unsubscribe = null;
    this.listeners.clear();
  }

  async refresh() {
    if (this.state.refreshing) return;
    this.state = { ...this.state, refreshing: true };
    this.emit();
    try {
      const items = await this.deps.refreshFn();
      const ts = this.now();
      this.state = { items: this.tagStale(items, ts), offline: false, refreshing: false, lastServerSyncAt: ts };
      this.emit();
    } catch (e) {
      // keep previous items, mark offline, and recompute staleness vs current time
      const retagged = this.tagStale(this.state.items, this.state.lastServerSyncAt);
      this.state = { ...this.state, items: retagged, refreshing: false, offline: true };
      this.emit();
    }
  }

  private tagStale(items: RideFeedItem[] | (RideFeedItem & { isStale?: boolean })[], lastServerSyncAt: number | null) {
    const now = this.now();
    const tooOld = lastServerSyncAt ? now - lastServerSyncAt > FIVE_MIN : true;
    return items.map((it: any) => ({ id: it.id, destinationCampus: it.destinationCampus, seatsAvailable: it.seatsAvailable, status: it.status, isStale: tooOld }));
  }

  private emit() {
    for (const l of this.listeners) l(this.state);
  }
}

// Default Firestore-powered subscribe/refresh
const defaultSubscribe: SubscribeToRideFeedFn = ({ onSnapshot: push, onError, geohashPrefixes }) => {
  const db = getFirestoreDb();
  let qRef: any;
  if (geohashPrefixes && geohashPrefixes.length === 1) {
    const prefix = geohashPrefixes[0];
    const start = prefix;
    const end = prefix + '\uf8ff';
    qRef = query(collection(db, 'ridePosts'), (global as any).where('status', '==', 'open'), (global as any).orderBy('geohash'), (global as any).startAt(start), (global as any).endAt(end));
  } else {
    qRef = query(collection(db, 'ridePosts'), (global as any).where('status', '==', 'open'));
  }
  const unsub = fsOnSnapshot(
    qRef,
    (snap: any) => {
      const now = Date.now();
      const items: RideFeedItem[] = snap.docs
        .map((d: any) => ({ id: d.id, ...(d.data() as any) }))
        .filter((d: any) => d.status === 'open' && (d.seatsAvailable ?? 0) > 0 && (!d.windowEnd?.toMillis || d.windowEnd.toMillis() > now))
        .map((d: any) => ({ id: d.id, destinationCampus: d.destinationCampus, seatsAvailable: d.seatsAvailable ?? 0, status: d.status }));
      const fromServer = !(snap.metadata?.fromCache ?? false);
      push(items, fromServer);
    },
    onError as any
  );
  return unsub;
};

const defaultRefresh: RefreshRideFeedFn = async () => {
  const db = getFirestoreDb();
  const q = query(collection(db, 'ridePosts'));
  const snap = await getDocs(q);
  return snap.docs
    .map((d: any) => ({ id: d.id, ...(d.data() as any) }))
    .filter((d: any) => d.status === 'open')
    .map((d: any) => ({ id: d.id, destinationCampus: d.destinationCampus, seatsAvailable: d.seatsAvailable ?? 0, status: d.status }));
};

export const useRideFeed = (
  deps: Partial<{ subscribe: SubscribeToRideFeedFn; refreshFn: RefreshRideFeedFn; now: () => number; geohashPrefixes: string[] }> = {}
) => {
  const controllerRef = useRef<RideFeedController | null>(null);
  if (!controllerRef.current) {
    controllerRef.current = new RideFeedController({
      subscribe: (opts) => (deps.subscribe ?? defaultSubscribe)({ ...opts, geohashPrefixes: deps.geohashPrefixes }),
      refreshFn: deps.refreshFn ?? defaultRefresh,
      now: deps.now
    });
  }
  const controller = controllerRef.current!;
  const [state, setState] = useState<RideFeedState>(controller.getState());

  useEffect(() => {
    const unsub = controller.subscribe(setState);
    return () => {
      unsub();
      controller.dispose();
      controllerRef.current = null;
    };
  }, [controller]);

  return useMemo(() => ({
    ...state,
    refresh: () => controller.refresh()
  }), [state, controller]);
};
