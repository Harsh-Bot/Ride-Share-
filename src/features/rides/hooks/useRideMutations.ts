import { useMutation } from '@tanstack/react-query';

export const useCreateLiveRide = () =>
  useMutation({
    mutationKey: ['create-live-ride'],
    mutationFn: async (_payload: unknown) => {
      // TODO: Implement ride creation via Firestore/Cloud Functions
      throw new Error('createLiveRide not implemented');
    }
  });

export const useCreateScheduledRide = () =>
  useMutation({
    mutationKey: ['create-scheduled-ride'],
    mutationFn: async (_payload: unknown) => {
      // TODO: Implement scheduled ride creation with validation
      throw new Error('createScheduledRide not implemented');
    }
  });

export const useRideStatusMutations = () =>
  useMutation({
    mutationKey: ['update-ride-status'],
    mutationFn: async (_payload: unknown) => {
      // TODO: Allow drivers to mark rides started/completed and trigger incentives
      throw new Error('updateRideStatus not implemented');
    }
  });
