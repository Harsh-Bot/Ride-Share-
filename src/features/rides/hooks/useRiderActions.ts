import { useEffect, useMemo, useRef, useState } from 'react';
import { FirebaseError } from 'firebase/app';
import { requestRide as realRequestRide, cancelRequest as realCancelRequest, cancelBooking as realCancelBooking } from '../api/requests';

type QueueItem =
  | { type: 'request'; args: [Parameters<typeof realRequestRide>[0]] }
  | { type: 'cancelRequest'; args: Parameters<typeof realCancelRequest> }
  | { type: 'cancelBooking'; args: Parameters<typeof realCancelBooking> };

export type RiderActionsState = {
  status: 'idle' | 'queued' | 'processing' | 'error';
  pendingCount: number;
  error: string | null;
};

type Deps = {
  requestRide: typeof realRequestRide;
  cancelRequest: typeof realCancelRequest;
  cancelBooking: typeof realCancelBooking;
};

export class RiderActionsController {
  private state: RiderActionsState = { status: 'idle', pendingCount: 0, error: null };
  private listeners = new Set<(s: RiderActionsState) => void>();
  private queue: QueueItem[] = [];

  constructor(private readonly deps: Deps) {}

  getState() { return this.state; }
  subscribe(listener: (s: RiderActionsState) => void) { this.listeners.add(listener); listener(this.state); return () => { this.listeners.delete(listener); }; }
  private emit() { for (const l of this.listeners) l(this.state); }

  private setError(error: FirebaseError | unknown, fallback: string) {
    const message = error instanceof FirebaseError ? error.message : fallback;
    this.state = { ...this.state, status: 'error', error: message };
    this.emit();
  }

  async requestRide(args: Parameters<typeof realRequestRide>[0]) {
    try {
      this.state = { ...this.state, status: 'processing', error: null };
      this.emit();
      await this.deps.requestRide(args);
      this.state = { ...this.state, status: this.queue.length ? 'queued' : 'idle' };
      this.emit();
    } catch (e) {
      const fe = e as FirebaseError & { message?: string };
      if (fe?.code === 'unavailable') {
        this.queue.push({ type: 'request', args: [args] });
        this.state = { status: 'queued', pendingCount: this.queue.length, error: null };
        this.emit();
        return;
      }
      if ((e as any)?.message === 'OUT_OF_RADIUS') {
        this.state = { ...this.state, status: 'error', error: "You're outside the ride zone" };
        this.emit();
        return;
      }
      this.setError(e, 'Action failed');
    }
  }

  async cancelRequest(...args: Parameters<typeof realCancelRequest>) {
    try {
      this.state = { ...this.state, status: 'processing', error: null };
      this.emit();
      await this.deps.cancelRequest(...args);
      this.state = { ...this.state, status: this.queue.length ? 'queued' : 'idle' };
      this.emit();
    } catch (e) {
      const fe = e as FirebaseError;
      if (fe?.code === 'unavailable') {
        this.queue.push({ type: 'cancelRequest', args });
        this.state = { status: 'queued', pendingCount: this.queue.length, error: null };
        this.emit();
        return;
      }
      this.setError(e, 'Action failed');
    }
  }

  async cancelBooking(...args: Parameters<typeof realCancelBooking>) {
    try {
      this.state = { ...this.state, status: 'processing', error: null };
      this.emit();
      await this.deps.cancelBooking(...args);
      this.state = { ...this.state, status: this.queue.length ? 'queued' : 'idle' };
      this.emit();
    } catch (e) {
      const fe = e as FirebaseError;
      if (fe?.code === 'unavailable') {
        this.queue.push({ type: 'cancelBooking', args });
        this.state = { status: 'queued', pendingCount: this.queue.length, error: null };
        this.emit();
        return;
      }
      this.setError(e, 'Action failed');
    }
  }

  async retryPending() {
    if (this.queue.length === 0) return;
    const next = this.queue.shift();
    this.state = { ...this.state, status: 'processing', pendingCount: this.queue.length };
    this.emit();
    if (!next) return;
    try {
      if (next.type === 'request') {
        await this.deps.requestRide(next.args[0] as any);
      } else if (next.type === 'cancelRequest') {
        const args = next.args as Parameters<typeof realCancelRequest>;
        await this.deps.cancelRequest(...args);
      } else if (next.type === 'cancelBooking') {
        const args = next.args as Parameters<typeof realCancelBooking>;
        await this.deps.cancelBooking(...args);
      }
      this.state = { ...this.state, status: this.queue.length ? 'queued' : 'idle' };
      this.emit();
    } catch (e) {
      const fe = e as FirebaseError;
      if (fe?.code === 'unavailable') {
        // push back to queue
        this.queue.unshift(next);
        this.state = { status: 'queued', pendingCount: this.queue.length, error: null };
        this.emit();
        return;
      }
      this.setError(e, 'Retry failed');
    }
  }
}

export const useRiderActions = (deps?: Partial<Deps>) => {
  const controllerRef = useRef<RiderActionsController | null>(null);
  if (!controllerRef.current) {
    controllerRef.current = new RiderActionsController({
      requestRide: deps?.requestRide ?? realRequestRide,
      cancelRequest: deps?.cancelRequest ?? realCancelRequest,
      cancelBooking: deps?.cancelBooking ?? realCancelBooking
    });
  }
  const controller = controllerRef.current!;
  const [state, setState] = useState<RiderActionsState>(controller.getState());
  useEffect(() => {
    const unsub = controller.subscribe(setState);
    return () => { unsub(); };
  }, [controller]);
  return useMemo(() => ({
    ...state,
    requestRide: (args: Parameters<typeof realRequestRide>[0]) => controller.requestRide(args),
    cancelRequest: (...args: Parameters<typeof realCancelRequest>) => controller.cancelRequest(...args),
    cancelBooking: (...args: Parameters<typeof realCancelBooking>) => controller.cancelBooking(...args),
    retryPending: () => controller.retryPending()
  }), [state, controller]);
};
