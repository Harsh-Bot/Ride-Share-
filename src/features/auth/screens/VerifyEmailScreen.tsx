import { useCallback, useMemo } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  KeyboardAvoidingView,
  Platform
} from 'react-native';
import * as Linking from 'expo-linking';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { AuthStackParamList } from '../../../navigation/types';
import { useAuth } from '../../../hooks/useAuth';
import { showToast } from '../../../utils/toast';

type Props = NativeStackScreenProps<AuthStackParamList, 'VerifyEmail'>;

const VerifyEmailScreen = ({ route }: Props) => {
  const { completeSignIn, isCompletingLink, pendingEmail, error } = useAuth();
  const email = useMemo(() => route.params?.email ?? pendingEmail ?? '', [route.params?.email, pendingEmail]);

  const handleRefreshSession = useCallback(async () => {
    try {
      const latestUrl = (await Linking.getInitialURL()) ?? '';
      if (!latestUrl) {
        showToast('Open the magic link from your inbox to finish signing in.', 'error');
        return;
      }
      await completeSignIn(latestUrl);
    } catch (attemptError) {
      showToast((attemptError as Error)?.message ?? 'We could not refresh your session.', 'error');
    }
  }, [completeSignIn]);

  return (
    <KeyboardAvoidingView
      style={styles.container}
      behavior={Platform.OS === 'ios' ? 'padding' : undefined}
      keyboardVerticalOffset={64}
    >
      <View style={styles.content}>
        <Text style={styles.title}>Check your inbox</Text>
        <Text style={styles.body}>
          We sent a magic link to {email ? <Text style={styles.emailText}>{email}</Text> : 'your email'}.
          {' '}Open it on this device to finish signing in. The link expires within 15 minutes for security.
        </Text>
        {error ? <Text style={styles.errorText}>{error}</Text> : null}
        <TouchableOpacity
          style={[styles.secondary, isCompletingLink ? styles.secondaryDisabled : null]}
          onPress={handleRefreshSession}
          disabled={isCompletingLink}
          accessibilityRole="button"
          accessibilityState={{ disabled: isCompletingLink }}
          testID="auth-refresh-session-button"
        >
          <Text style={styles.secondaryText}>
            {isCompletingLink ? 'Looking for your linkâ€¦' : 'I already tapped the link'}
          </Text>
        </TouchableOpacity>
      </View>
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFFFFF'
  },
  content: {
    flex: 1,
    padding: 24,
    justifyContent: 'center'
  },
  title: {
    fontSize: 28,
    fontWeight: '700',
    marginBottom: 12
  },
  body: {
    fontSize: 16,
    marginBottom: 24
  },
  emailText: {
    fontWeight: '600'
  },
  errorText: {
    color: '#C62828',
    marginBottom: 12
  },
  secondary: {
    padding: 16,
    borderRadius: 12,
    alignItems: 'center'
  },
  ctaDisabled: {
    opacity: 0.6
  },
  ctaText: {
    color: '#FFFFFF',
    fontWeight: '600'
  },
  secondaryDisabled: {
    opacity: 0.6
  }
});

export default VerifyEmailScreen;
