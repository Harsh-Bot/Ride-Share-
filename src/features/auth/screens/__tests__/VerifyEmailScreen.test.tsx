import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import VerifyEmailScreen from '../VerifyEmailScreen';
import { useAuth } from '../../../../hooks/useAuth';

type MockReturn = {
  completeSignIn: jest.Mock;
  authError: string | null;
  isLoading: boolean;
  pendingEmail: string | null;
};

jest.mock('../../../../hooks/useAuth');

const mockCompleteSignIn = jest.fn();
const authMock: MockReturn = {
  completeSignIn: mockCompleteSignIn,
  authError: null,
  isLoading: false,
  pendingEmail: 'student@sfu.ca'
};

(useAuth as jest.Mock).mockImplementation(() => authMock);

describe('VerifyEmailScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    authMock.authError = null;
    authMock.isLoading = false;
    authMock.pendingEmail = 'student@sfu.ca';
  });

  it('matches snapshot', () => {
    const tree = render(
      <VerifyEmailScreen
        navigation={undefined as never}
        route={{ key: 'verify', name: 'VerifyEmail', params: { email: 'student@sfu.ca' } }}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('calls completeSignIn when six digit code entered', () => {
    const { getByPlaceholderText, getByText } = render(
      <VerifyEmailScreen
        navigation={undefined as never}
        route={{ key: 'verify', name: 'VerifyEmail', params: { email: 'student@sfu.ca' } }}
      />
    );

    fireEvent.changeText(getByPlaceholderText('000000'), '123456');
    fireEvent.press(getByText('Verify and continue'));

    expect(mockCompleteSignIn).toHaveBeenCalledWith('123456');
  });
});
