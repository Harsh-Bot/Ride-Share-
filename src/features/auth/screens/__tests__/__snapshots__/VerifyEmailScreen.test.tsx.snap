// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmailScreen matches snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "center",
      "padding": 24,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 28,
        "fontWeight": "700",
        "marginBottom": 12,
      }
    }
  >
    Enter your verification code
  </Text>
  <Text
    style={
      {
        "fontSize": 16,
        "marginBottom": 24,
      }
    }
  >
    We sent a 6-digit code to 
    student@sfu.ca
    . Enter it below to continue.
  </Text>
  <TextInput
    keyboardType="number-pad"
    maxLength={6}
    onChangeText={[Function]}
    placeholder="000000"
    style={
      {
        "borderColor": "#CCCCCC",
        "borderRadius": 12,
        "borderWidth": 1,
        "fontSize": 20,
        "letterSpacing": 8,
        "marginBottom": 16,
        "padding": 16,
        "textAlign": "center",
      }
    }
    textContentType="oneTimeCode"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#D4145A",
        "borderRadius": 12,
        "opacity": 0.6,
        "padding": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontWeight": "600",
        }
      }
    >
      Verify and continue
    </Text>
  </View>
</View>
`;
